### DISCUSSION TITLE
Make PasswordPolicyProvider SPI more flexible

### DISCUSSION THREAD

**Discussion by thomasdarimont** [keycloak/keycloak#26517] [Ideas]
# Overview
Currently, Keycloak supports a decent set of password policy rules that can be used to compose complex password policies on the realm level. However, in many projects, an often raised requirement is the support for multiple password policies, which can be applied to a subset of users, e.g., for all members of a group or user with a specific attribute.

Such requirements are currently challenging to implement because the password policy API and implementations implicitly load their configuration from the `PasswordPolicy` obtained from the realm.
An example of this case can be seen in the [validate method of the LengthPasswordPolicyProvider](
https://github.com/keycloak/keycloak/blob/main/server-spi-private/src/main/java/org/keycloak/policy/LengthPasswordPolicyProvider.java#L39).
Since the configuration is fetched from the realm, feeding a context-specific password policy into the password validation is challenging.

# Design
If we would slightly change the password policy SPI, e.g., by allowing to pass in the `PasswordPolicy` directly to the `validate` method, we could effectively have password policies that are computed in a context-specific way, e.g., based on a group/tenant attribute, user attribute, etc. 

This approach would allow the users to declare flexible password policies that could be derived dynamically.

# Goals
- Reduce coupling of PasswordPolicyProvider configuration access from Realms
- Ease testing of PasswordPolicyProvider implementations with different configurations
- Enable password policy evaluations based on a given password policy
- Enable construction of password policies based on different sources, e.g. Group / Role attributes and other sources. 
 
# Possible Solution

In my branch [refactor-password-policy](https://github.com/thomasdarimont/keycloak/tree/poc/refactor-password-policy), I experimented with this approach, and I think this could be implemented with just a few minor changes in a backward compatible way.
In my PoC I introduced a [`PasswordPolicyContext`](https://github.com/thomasdarimont/keycloak/blob/poc/refactor-password-policy/server-spi/src/main/java/org/keycloak/policy/PasswordPolicyContext.java#L26) which holds the data necessary for password policy validation. 
This `PasswordPolicyContext` is now passed to a new method 
`PolicyError validate(String password, PasswordPolicyContext policyContext);` which I added to the `PasswordPolicyManagerProvider` and `PasswordPolicyProvider` interfaces and respective implementations, e.g. [DefaultPasswordPolicyMnaagerProvider](https://github.com/thomasdarimont/keycloak/blob/poc/refactor-password-policy/server-spi-private/src/main/java/org/keycloak/policy/DefaultPasswordPolicyManagerProvider.java#L51).

I adapted the internal password validation code to delegate to the new `PolicyError validate(String password, PasswordPolicyContext policyContext);` method. Here is an example for the [LengthPasswordPolicyProvider](https://github.com/thomasdarimont/keycloak/blob/poc/refactor-password-policy/server-spi-private/src/main/java/org/keycloak/policy/LengthPasswordPolicyProvider.java#L49).

This change now allows users to perform password policy validation with a custom `PasswordPolicyContext`, which can be populated with an explicit `PasswordPolicy` derived from some context, e.g., a group attribute, etc.

IMHO, this provides enough flexibility for context-specific password policy definitions and validations.

As a positive side effect, this eases testing of `PasswordPolicyProvider` implementations with different configurations since we can pass in a custom `PasswordPolicy` via the `PasswordPolicyContext` directly.

If you think this makes sense, I'm more than happy to send a PR :)

**Comment by jbman** [comment]
Especially when password policies vary among users, it would be good if each policy can also provide (a configuration dependent) password requirement text in addition to the error text. This should be used to render password requirements to the user on registration or also to the admin who may set a password.

**Comment by dp-carrier** [comment]
Did anything ever come of this idea?
